src/python/pants/backend/scala/compile/scalac_plugins.py

```diff
@@ -99,6 +113,65 @@ async def add_resolve_name_to_plugin_request(
     )
 
 
+@dataclass(frozen=True)
+class _ResolveScalacPluginArtifact:
+    field: ScalacPluginArtifactField
+    consumer_target: Target
+
+
+@rule
+async def _resolve_scalac_plugin_artifact(
+    request: _ResolveScalacPluginArtifact,
+    target_types_to_generate_requests: TargetTypesToGenerateTargetsRequests,
+    local_environment_name: ChosenLocalEnvironmentName,
+    field_defaults: FieldDefaults,
+) -> WrappedTarget:
+    """Helps resolving the actual artifact for a scalac plugin even in the scenario in which the
+    artifact has been declated as a scala_artifact and it has been parametrized (i.e. across
+    multiple resolves for cross building)."""
+
+    environment_name = local_environment_name.val
+
+    address = await Get(Address, AddressInput, request.field.to_address_input())
+
+    parametrizations = await Get(
+        _TargetParametrizations,
+        {
+            _TargetParametrizationsRequest(
+                address.maybe_convert_to_target_generator(),
+                description_of_origin=(
+                    f"the target generator {address.maybe_convert_to_target_generator()}"
+                ),
+            ): _TargetParametrizationsRequest,
+            environment_name: EnvironmentName,
+        },
+    )
+
+    target = parametrizations.get_subset(
+        address, request.consumer_target, field_defaults, target_types_to_generate_requests
+    )
```

Comment:
Yea, as you've mentioned, `_TargetParametrizations` is a very awkward API, which is part of why it is still explicitly private. I'd like to find a cleaner way to expose the `get_subset` call... it may be that it is a good use-case for one of our first public rule helpers (rather than exposing the `_TargetParametrizations` type, which should maybe be an implementation detail).

`get_subset` is the fundamental operation, but it's possible that there is a better granularity to expose, such as one level up in `resolve_dependencies`: it's also possible that exposing [`resolve_dependencies`](https://github.com/pantsbuild/pants/blob/5becc13ebaad5ecdc42228486727da1c20521521/src/python/pants/engine/internals/graph.py#L1287-L1295) as an API that can be used to expand any given list of dependencies would make sense?

Reply:
In fact I did come up with this implementation by looking at the implementation of the `resolve_dependencies` rule in the graph. However getting the subset is only one step in the process, that results in getting the actual target generator, we still need the final generated target and it's why there is a call to `parametrizations.generated_for` right after.

Although exposing the subset will get a long way, because IIRC there is a dependency relationship between a target generator and its generated targets, so I guess I could resolve the dependencies of the generator and, in case I find none or more than one, then error there.

In a way, the field `artifact` in `scalac_plugin` is like a `SingleDependencyField` for which we expect to find only one target after expanding parameters and target generation. What makes it even more awkward is the fact that the _consumer target_, the one that carries the parameters over, is not the `scalac_plugin` itself, but the `scala_sources` target that lists the plugin names in the `scalar_plugins` field (a field that gives me `SpecialCasedDependencies` vibes, but that's a different topic).

So, in a way, everything here smells of this being _a special case_:
1. A target (`scalac_plugin`) that adds metadata to the build and it's meant to _decorate_ another target (`jvm_artifact`/`scala_artifact`).
1. A 1:1 target generator (`scala_artifact`->`jvm_artifact`)
1. Parametrized fields
1. A _compile-only dependency_ between a `scala_source(s)` target and the `jvm_artifact` target generated, and the dependency happens via an intermediary target (the `scalac_plugin`)
1. Also, Scalac plugins can be globally defined in the `scalac` subsystem using a `dict` where the key is the resolve name.

In general, this is hard to generalise as I can only the Kotlin backend implements compiler plugins in a similar fashion and it's unlikely this scenario shows up in there too.

Sorry for the write up!

Reply:
As a side note, there is another case that I'm aware of where having a field for a single dependency would be nice: The `helm_deployment` target. That target needs a reference to the Helm chart to deploy and there should be only one of those.

At first I wanted to model using a bespoke field but ended up asking users to add the address to the chart in the `dependencies` field as I wanted Pants to rebuild in code that may have changed. While this works, it feels unpolished and the implementation is a bit awkward as it has to avoid a rule cycle so I always wanted to change that and in fact I have a branch somewhere with some changes.